src/libtcod/renderer_xterm.c

```diff
@@ -159,7 +175,257 @@ static void xterm_destructor(struct TCOD_Context* __restrict self) {
+          break;
+        case 'Q':
+          send_sdl_key_press(SDLK_F2, false);
+          break;
+        case 'R':
+          if (g_waiting_for_get_size) {
+            g_got_rows = arg0;
+            g_got_columns = arg1;
+            g_got_size_timestamp = SDL_GetTicks();
+          } else {
+            send_sdl_key_press(SDLK_F3, false);
+          }
+          break;
+        case 'S':
+          send_sdl_key_press(SDLK_F4, false);
+          break;
+        case '~':
+          switch (arg0) {
+            case 11:
+              send_sdl_key_press(SDLK_F1, false);
+              break;
+            case 12:
+              send_sdl_key_press(SDLK_F2, false);
+              break;
+            case 13:
+              send_sdl_key_press(SDLK_F3, false);
+              break;
+            case 14:
+              send_sdl_key_press(SDLK_F4, false);
+              break;
+            case 15:
+              send_sdl_key_press(SDLK_F5, false);
+              break;
+            case 17:
+              send_sdl_key_press(SDLK_F6, false);
+              break;
+            case 18:
+              send_sdl_key_press(SDLK_F7, false);
+              break;
+            case 19:
+              send_sdl_key_press(SDLK_F8, false);
+              break;
+            case 20:
+              send_sdl_key_press(SDLK_F9, false);
+              break;
+            case 21:
+              send_sdl_key_press(SDLK_F10, false);
+              break;
+            case 23:
+              send_sdl_key_press(SDLK_F11, false);
+              break;
+            case 24:
+              send_sdl_key_press(SDLK_F12, false);
+          }
+          break;
+      }
+      break;
+    case 'O': // SS3
+      switch (end) {
+        case 'P':
+          send_sdl_key_press(SDLK_F1, false);
+          break;
+        case 'Q':
+          send_sdl_key_press(SDLK_F2, false);
+          break;
+        case 'R':
+          send_sdl_key_press(SDLK_F3, false);
+          break;
+        case 'S':
+          send_sdl_key_press(SDLK_F4, false);
+          break;
+      }
+      break;
+  }
+}
+
+static int xterm_handle_input(void *arg) {
+  while (true) {
+    int ch = getchar();
+    if (ch == '\x1b') {
+      xterm_handle_input_escape();
+      continue;
+    }
+    send_sdl_key_press(ch, isupper(ch));
+  }
+  return 0;
+}
+
+static TCOD_Error xterm_recommended_console_size(
+    struct TCOD_Context* __restrict self,
+    float magnification,
+    int* __restrict columns,
+    int* __restrict rows) {
+  fprintf(stdout, "\x1b[%i;%iH", INT_MAX, INT_MAX);
+  fflush(stdout);
+  g_waiting_for_get_size = true;
+  Uint32 start_time = SDL_GetTicks();
+  fprintf(stdout, "\x1b[6n");
+  fflush(stdout);
+  while (!SDL_TICKS_PASSED(SDL_GetTicks(), start_time + 100)) {
+    if (SDL_TICKS_PASSED(g_got_size_timestamp, start_time)) {
+      *columns = g_got_columns;
+      *rows = g_got_rows;
+      g_waiting_for_get_size = false;
+      return TCOD_E_OK;
+    }
+  }
```

Comment:
I'd prefer that this wasn't a busy wait.  Could you add `SDL_Delay(1);` here?
```suggestion
  while (!SDL_TICKS_PASSED(SDL_GetTicks(), start_time + 100)) {
    if (SDL_TICKS_PASSED(g_got_size_timestamp, start_time)) {
      *columns = g_got_columns;
      *rows = g_got_rows;
      g_waiting_for_get_size = false;
      return TCOD_E_OK;
    }
    SDL_Delay(1);
  }
```

Reply:
That should be fine. The timeout is another thing I don't really like, but it is so the thread doesn't completely hang in the case where terminal doesn't support the get cursor position request or otherwise fails to send what the code expects.